### PR TITLE
Here's a numpy example from a numpy tutorial on doing AQI

### PR DESCRIPTION
**Pull request by skrawcz**
See https://github.com/numpy/numpy-tutorials/tree/main/content
for where this is pulled from.

The names of the functions could be better I think. Otherwise this
doesn't seem like a bad pipeline. Versus doing this analysis in a notebook,
the strength of Hamilton here is in forcing concise definitions and language
around steps in the analysis -- and then magically the analysis is pretty
reusable / very easy to augment.

E.g. add some @config.when or split things into python modules to be swapped out.

## Additions

- example showing using pure numpy code.

## Removals

- N/A

## Changes

- N/A

## Testing

1. Script runs with `python run_analysis.py`

## Notes

- This is based on https://github.com/numpy/numpy-tutorials/tree/main/content

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the [TODO link to standards]()
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python

~- [ ] python 3.6~
- [x] python 3.7

### CODE REVIEW

File: examples/numpy/air-quality-analysis/analysis_flow.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    :return: index computed
+    """
+    bp = breakpoints[pol]
+
+    if pol == 'CO':
+        inc = 0.1
+    else:
+        inc = 1
+
+    if bp[0] <= con < bp[1]:
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/analysis_flow.py:40]: This is copied I imagine? There's a cleaner way to write this code...

File: examples/numpy/air-quality-analysis/analysis_flow.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+# --- private helper functions
+def _moving_mean(a, n):
+    """Computes the moving mean using numpy constructurs."""
+    ret = np.cumsum(a, dtype=float, axis=0)
+    ret[n:] = ret[n:] - ret[:-n]
+    return ret[n - 1:] / n
+
+
+def _compute_indices(breakpoints: dict, aqi: np.ndarray, pol: str, con: float) -> float:
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/analysis_flow.py:24]: Type-annotations for breakpoints might be nice

File: examples/numpy/air-quality-analysis/analysis_flow.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+
+def pollutants_A_24hr_avg(pollutants_A: np.ndarray) -> np.ndarray:
+    """24 hour move average of pollutant A"""
+    return _moving_mean(pollutants_A, 24)
+
+
+def pollutants_B_8hr_avg(pollutants_B: np.ndarray, pollutants_A_24hr_avg: np.ndarray) -> np.ndarray:
+    """8 hour move average of pollutant B"""
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/analysis_flow.py:120]: Why does this take in `pollutants_A_24hr_avg`? For the shape? Might be nice to have more descriptive docstrings, or utilize an intermediate step that computes the shape.

File: examples/numpy/air-quality-analysis/run_analysis.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+"""
+Run this analysis by doing `python run_analysis.py`.
+
+Or copy this code into a notebook for further analysis...
+
+Otherwise this is meant to mirror analysis as presented in https://github.com/numpy/numpy-tutorials/tree/main/content.
+"""
+
+if __name__ == '__main__':
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/run_analysis.py:13]: This all looks pretty clean -- maybe include a graph view/image as part of the tutorial?

File: examples/numpy/air-quality-analysis/run_analysis.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    }, analysis_flow, adapter=adapter)
+
+    output = ['t_value', 'p_value', 'before_sample', 'after_sample']
+    # dr.visualize_execution(output, './my_file.dot', {})
+    result = dr.execute(output)
+    print(result)
+    print(f"The t value is {result['t_value']} and the p value is {result['p_value']}.")
+
+    # from a dict result, it's easy to use another ResultMixin to build another result
+    sample_matrix = base.NumpyMatrixResult().build_result(before_sample=result['before_sample'],
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/run_analysis.py:30]: Not exactly sure why this is here? Is the point of a `ResultsMixin` to be called by the framework, called by the user, or both?

File: examples/numpy/air-quality-analysis/analysis_flow.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        'PM10': np.array([0, 51, 101, 251, 351, 431]),
+        'NO2': np.array([0, 41, 81, 181, 281, 401]),
+        'NH3': np.array([0, 201, 401, 801, 1201, 1801]),
+        'SO2': np.array([0, 41, 81, 381, 801, 1601]),
+        'CO': np.array([0, 1.1, 2.1, 10.1, 17.1, 35]),
+        'O3': np.array([0, 51, 101, 169, 209, 749])
+    }
+
+
+def pollutants_A_24hr_avg(pollutants_A: np.ndarray) -> np.ndarray:
```
Comment (elijahbenizzy) [inline examples/numpy/air-quality-analysis/analysis_flow.py:114]: IMO this demonstrates where pandas would be much cleaner -- dt indices associated with it make joining/rolling averages much safer.

**Review by elijahbenizzy — COMMENTED** [review]
Nice. I think this is actually a great use for validators on the shape of the data, given that typing is just `np.ndarray`

File: examples/numpy/air-quality-analysis/analysis_flow.py (hunk 6 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/analysis_flow.py:114]: yep, they do say that in the post at the end too.

File: examples/numpy/air-quality-analysis/run_analysis.py (hunk 5 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/run_analysis.py:30]: both ... just trying to show functionality.

File: examples/numpy/air-quality-analysis/analysis_flow.py (hunk 1 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/analysis_flow.py:40]: yep this is all copied.

File: examples/numpy/air-quality-analysis/analysis_flow.py (hunk 3 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/analysis_flow.py:120]: /shrug -- i'll see what the notebook says.

File: examples/numpy/air-quality-analysis/analysis_flow.py (hunk 2 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/analysis_flow.py:31]: we don't support generics yet.

File: examples/numpy/air-quality-analysis/run_analysis.py (hunk 4 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/run_analysis.py:15]: done.

File: examples/numpy/air-quality-analysis/run_analysis.py (hunk 5 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/run_analysis.py:33]: added a comment.

File: examples/numpy/air-quality-analysis/analysis_flow.py (hunk 3 above)
Comment (skrawcz) [reply examples/numpy/air-quality-analysis/analysis_flow.py]: added.

**Comment by elijahbenizzy** [comment]
Adding some end-to-end tests could be nice -- this could be an integration test

**Review by elijahbenizzy — APPROVED** [review]
LGTM!

**Pull request merged**